serial-multiplexer/src/main.rs

```diff
@@ -42,130 +54,139 @@ fn main() {
+            let (mut master, slave) = TTYPort::pair().expect("Unable to create ptty pair");
+            master.set_timeout(Duration::MAX).unwrap();
+
+            let name = slave.name().unwrap();
+            unused.push(slave);
+
+            let mut link_path = std::env::home_dir().unwrap_or(PathBuf::from("/dev"));
+
+            link_path.push("vtty");
+            if !link_path.exists() {
+                create_dir(&link_path).unwrap();
+            }
+
+            link_path.push(f.0);
+            let _ = remove_file(&link_path);
+
+            symlink(name, link_path).unwrap();
+
+            let serial_port_manager = SerialPortManager::with_port(master);
+            let serial_port_manager_ref = Arc::new(Mutex::new(serial_port_manager));
+
+            sender_processors.push(SerialConnectionSenderProcessor {
+                id: entry.id,
+                port_manager: serial_port_manager_ref.clone(),
+                port_receiver: port_receiver,
+            });
 
-        communicate(&mut multiplexed_port, &mut serial_ports);
+            receiver_processors.push(SerialConnectionReceiverProcessor {
+                id: entry.id,
+                port_manager: serial_port_manager_ref,
+                write_to_main_bus: main_bus_sender.clone(),
+            });
+
+            senders.push(SerialConnectionSender {
+                id: entry.id,
+                port_sender,
+            });
+        });
     }
+
+    println!("Starting communication loop...");
+    communicate(
+        sender_processors,
+        receiver_processors,
+        senders,
+        main_bus_receiver,
+        &mut multiplexed_port,
+    );
 }
 
 fn communicate(
+    sender_processors: Vec<SerialConnectionSenderProcessor>,
+    receiver_processors: Vec<SerialConnectionReceiverProcessor>,
+    senders: Vec<SerialConnectionSender>,
+    main_bus_receiver: Receiver<DataBlock>,
     multiplexed_port: &mut TTYPort,
-    serial_ports: &mut HashMap<u32, SerialEntry>,
 ) {
-    let mut multiplexed_port_clone = multiplexed_port.try_clone_native().unwrap();
+    sender_processors.into_iter().for_each(|f| {
+        std::thread::spawn(move || {
+            f.process_loop();
+        });
+    });
 
-    let mut serial_ports_clone: HashMap<u32, SerialEntry> = HashMap::new();
-
-    serial_ports.iter().for_each(|port| {
-        serial_ports_clone.insert(
-            port.0.clone(),
-            SerialEntry { 
-                name: port.1.name.clone(), 
-                device: port.1.device.try_clone_native().unwrap(), 
-                id: port.1.id
-            },
-        );
+    receiver_processors.into_iter().for_each(|f| {
+        std::thread::spawn(move || {
+            f.process_loop();
+        });
     });
 
-    std::thread::spawn(move || {
-        let mut local_map = serial_ports_clone;
-        
+    let mut multiplexed_port_clone = multiplexed_port.try_clone_native().unwrap();
 
+    std::thread::spawn(move || {
         loop {
```

Comment:
Can we get signal hooks? Like, at least listening on SIGINT?

Reply:
@TheDevMinerTV Just double checking, is this needed? Can you link me to docs/etc explaining why this is needed? (Or would i just need a signal handler in the main thread? Even then, what would i need to handle?)